go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +631,562 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+	//		//if errors.Is(err, context.Canceled) ||
+	//		//	errors.Is(err, context.DeadlineExceeded) {
+	//		//	newState = vcopierCopyTaskCancel
+	//		//} else {
+	//		//	newState = vcopierCopyTaskFail
+	//		//}
+	//	}
+	//}()
+
+	newState = nextState
+
+	// Checking for context errors seems like a sensible thing to do. E.g. when
+	// copyPhaseDuration elapses it will cause the task to be canceled. Hower
+	// this is more aggressive cancelation behavior than the unit tests
+	// currently support. Will need to rework unit tests before we enable this.
+
+	//if err = ctx.Err(); err != nil {
+	//	return
+	//}
+	if err = vtl.Before(nextState).Notify(ctx, args); err != nil {
+		newState = vcopierCopyTaskFail
+		return
+	}
+	if err = fn(ctx, args); err != nil {
+		newState = vcopierCopyTaskFail
+		return
+	}
+	if err = vtl.After(nextState).Notify(ctx, args); err != nil {
+		newState = vcopierCopyTaskFail
+		return
+	}
+	return
+}
+
+// Do registers a callback with the vcopierCopyTaskResultHooks, to be triggered
+// when a task advances to a "done" state (= canceled, completed, failed).
+func (vrh *vcopierCopyTaskResultHooks) Do(fn func(context.Context, *vcopierCopyTaskResult)) {
+	vrh.fns = append(vrh.fns, fn)
+}
+
+// Notify triggers all callbacks registered with this vcopierCopyTaskResultHooks.
+func (vrh *vcopierCopyTaskResultHooks) Notify(ctx context.Context, result *vcopierCopyTaskResult) {
+	for _, fn := range vrh.fns {
+		fn(ctx, result)
+	}
+}
+
+// SendTo registers a hook that accepts a result and sends the result to the
+// provided channel. E.g.:
+//
+//	resultCh := make(chan *vcopierCopyTaskResult, 1)
+//	task.Lifecycle().OnResult().SendTo(resultCh)
+//	defer func() {
+//	  result := <-resultCh
+//	}()
+func (vrh *vcopierCopyTaskResultHooks) SendTo(ch chan<- *vcopierCopyTaskResult) {
+	vrh.Do(func(ctx context.Context, result *vcopierCopyTaskResult) {
+		select {
+		case ch <- result:
+		case <-ctx.Done():
+			// Log a warning?
+		}
+	})
+}
+
+// AwaitCompletion registers a callback that returns an error unless the
+// provided chan produces a vcopierTaskResult in a complete state.
+//
+// This is useful for sequencing vcopierCopyTasks, e.g.:
+//
+//	resultCh := make(chan *vcopierCopyTaskResult, 1)
+//	prevT.Lifecycle().OnResult().SendTo(resultCh)
+//	currT.lifecycle().Before(vcopierCopyTaskUpdateCopyState).AwaitCompletion(resultCh)
+func (vth *vcopierCopyTaskHooks) AwaitCompletion(resultCh <-chan *vcopierCopyTaskResult) {
+	vth.Do(func(ctx context.Context, args *vcopierCopyTaskArgs) error {
+		select {
+		case result := <-resultCh:
+			if result == nil {
+				return fmt.Errorf("channel was closed before a result received")
+			}
+			if !result.State().IsDone() {
+				return fmt.Errorf("received result is not done")
+			}
+			if result.State() != vcopierCopyTaskComplete {
+				return fmt.Errorf("received result is not complete")
+			}
+			return nil
+		case <-ctx.Done():
+			// Arguably we should return ctx.Err() here.
+			return nil
+		}
```

Comment:
Again, here this seems like a potentially serious failure, or no? I would expect this to be treated as an error but we don't seem to be doing any error handling here or on the caller side. Similar questions as above for `SendTo()`.

Reply:
We totally can and should enable this. We will just need to re-worker the tests in `vcopier_test.go`. Will talk through that when we sync.

Reply:
I could have sworn last time I returned `ctx.Err()` from here it broke tests, but just added it back and now it's working. Yay shrug.